libraries/common/list.effekt

```diff
@@ -167,9 +167,52 @@ def foreachIndex[A](list: List[A]){ f: (Int, A) {Control} => Unit }: Unit = {
 /// Map a function `f` over elements in a given list.
 ///
 /// O(N)
-def map[A, B](l: List[A]) { f: A => B } : List[B] = {
-  var acc = Nil[B]()
-  l.foreach { el => acc = Cons(f(el), acc) }
+def map[A, B](l: List[A]) { f: A => B }: List[B] = {
+  var acc: List[B] = Nil()
+  l.foreach { el =>
+    acc = Cons(f(el), acc)
+  }
+  acc.reverse
+}
+
+/// Map a function `f` over elements and their indices in a given list.
+///
+/// O(N)
+def mapWithIndex[A, B](l: List[A]) { f: (Int, A) => B }: List[B] = {
+  var acc: List[B] = Nil()
+  l.foreachIndex { (i, el) =>
+    acc = Cons(f(i, el), acc)
+  }
+  acc.reverse
+}
+
+/// Map a function `f` over elements in a given list,
+/// keeping only the elements for which the function returned `true`,
+/// discarding the elements for which the function returned `false`.
+///
+/// O(N)
+def filter[A](l: List[A]) { shouldKeep: A => Bool }: List[A] = {
+  var acc: List[A] = Nil()
+  l.foreach { a =>
+    if (shouldKeep(a)) {
+      acc = Cons(a, acc)
+    }
+  }
+  acc.reverse
+}
+
+/// Map a function `f` over elements and their indices in a given list,
+/// keeping only the elements for which the function returned `true`,
+/// discarding the elements for which the function returned `false`.
+///
+/// O(N)
+def filterWithIndex[A](l: List[A]) { shouldKeep: (Int, A) => Bool }: List[A] = {
+  var acc: List[A] = Nil()
+  l.foreachIndex { (i, a) =>
+    if (shouldKeep(i, a)) {
+      acc = Cons(a, acc)
+    }
+  }
   acc.reverse
 }
```

Comment:
Same for these. This can all be expressed in terms of `foldRight` I think. Though, I think this is a matter of preference and not too important.

Reply:
dtto above
I don't care about the style here (functional v imperative) too much, I'm mostly mirroring the existing code.

Reply:
I disagree that this is about functional vs imperative. This is about duplication of existing code and DRY. But if you agree that this is fine here, I respect that.

Reply:
> For me, this is about duplication of existing code and DRY. 

I'd like the code to be DRYer as well, yes, but last I've tried to define everything through, let's say, `collect`, the inliner couldn't keep up in a bigger app (if a `map` is defined via `collect`, then I'd want it to desugar to a `foreach[Index]` in JS every single time).

There are already ensions about the stdlib being very slow and bloated, I'd hate to make the problem worse when adding a few functions used by the students.

Reply:
Sorry to chime into the bike shed, but for me this is indeed a functional vs imperative debate. 

And I must say, I always have problems reading foldX code (I can never remember the order of traversal) and strictly prefer the imperative loop.